### PR TITLE
Live Camera Feed Enhancement(Review, QA and testing)

### LINKED ISSUE #7: [C4GT] Live Camera Feed Enhancement
## Project Detail

CARE is a centralized capacity management and patient management system, central to the 10BedICU Project, integrating patients, doctors, hospitals, labs, specialized treatment centers, hospital administrators, and shifting control cells. Hospitals update crucial information about their assets, providing district administration with a comprehensive view of the healthcare system via smart dashboards. CARE digitizes patient records, streamlines workflows for pandemic management, and is deployed in remote areas, enabling TeleICU services for underserved citizens. It revolutionizes healthcare management, enhancing efficiency, accessibility, and patient outcomes.

## Features To Be Implemented

This proposal outlines the modifications to the existing Live Camera Feed feature used in healthcare settings for real-time patient diagnosis. The proposed enhancements aim to provide doctors with improved control and collaborative capabilities while viewing patients remotely.

- **Boundary-Based Camera Visual**: Implement a pre-fixed boundary on the XY axis to limit camera movement per patient, ensuring focused viewing within specific areas for accurate diagnosis.
- **User Notification for Concurrent Viewers**: Notify users when another viewer is already accessing the camera feed for a patient, promoting better coordination and reducing interruptions.
- **Request Camera Toggle Access**: Allow users to request temporary camera access from the current viewer, fostering collaboration and efficient resource utilization.
- **Privacy Control Alert**: If a patient has privacy control enabled, doctors will receive an alert indicating the privacy control status. They will be advised to contact hospital staff for assistance in removing the privacy control and once removed only, the doctor should be able to access the patient's camera feed.

## Learning Path

|  | Details |
| :---:  | :---:  |
| Complexity       | Medium |
| Required Skills | ReactJS, NodeJS, TypeScript, Django |
| Mentors            | @nihal467 , @vigneshhari , @khavinshankar , @rithviknishad , @gigincg , @mathew-alex   |
| Project Size      | 6 Weeks |

## Product Set-up

To set up the project on a system, follow these guidelines:

- Install the required programming language and libraries, such as Python and their respective data validation libraries.
- Fork the project
- Clone the forked project repository from GitHub.
- Set up the environment variables and configurations required to run the utility. (Refer: https://care-be-docs.coronasafe.network/local-setup/configuration.html)

For more details, please refer to the readme file available at: https://github.com/coronasafe/care
_For any issues with setting up camera, get in touch with @mathew-alex_ 

## Acceptance Criteria

- Camera Feed: Smooth operation video streaming on the patient consultation page.
- Privacy Mode Alerts: Seamless alerts for privacy mode with timely notifications to doctors.
- Toggle Access: Proper functionality and smooth switching of camera access.
- QA and Code Review: Thorough testing and comprehensive code review to ensure compliance and quality.

## Milestone

- Product Understanding: Gain understanding of Care product workflow.
- Middleware and Camera Workflow: Understand middleware and camera operations.
- Privacy Mode and Boundary-Based Feature Implementation: Implement privacy mode and boundary-based features in the camera.
- QA and Code Review: Conduct QA testing and code review.

### PR DESCRIPTION
**Pull request by JahnabDutta**
### <samp>🤖 Generated by Copilot at da93272</samp>

This pull request adds new features and improvements to the camera feed and the notification system in the frontend application. It enables users to add, update, delete, and preview camera boundary presets for beds, as well as toggle patient privacy mode for the camera feed. It also refactors the notification subscription logic into a custom hook and adds end-to-end tests for the camera features. It modifies several components, hooks, models, constants, actions, and routes to implement these features and improvements. It also fixes some minor issues with the code formatting and style.

## Proposed Changes

- Testing and QA related PR for Live Camera Feed Enhacnement feature.

@coronasafe/care-fe-code-reviewers @coronasafe/code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.coronasafe.network/coronasafe-care-documentation/architecture/architecture-and-layering-of-care).
- [ ] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [ ] Request for Peer Reviews
- [ ] Completion of QA

### <samp>🤖 Generated by Copilot at da93272</samp>

*  Add and implement custom hook `useNotificationSubscribe` for managing notification subscription state and actions (F4, [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-75a5f730c112dc5dbfb5da454ea076adc2632003c0372b04ddc9be5752c4e586L4-R9), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-75a5f730c112dc5dbfb5da454ea076adc2632003c0372b04ddc9be5752c4e586L26-R19), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-75a5f730c112dc5dbfb5da454ea076adc2632003c0372b04ddc9be5752c4e586L155), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-75a5f730c112dc5dbfb5da454ea076adc2632003c0372b04ddc9be5752c4e586L166-R164), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-75a5f730c112dc5dbfb5da454ea076adc2632003c0372b04ddc9be5752c4e586L180-L205), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-75a5f730c112dc5dbfb5da454ea076adc2632003c0372b04ddc9be5752c4e586L234-L308))
* Add and implement custom hook `useFeedPTZ` for managing camera feed and asset operations ([link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-1e5a8875fbda16cf406f25e6def8d955c7774eeb32da266d24185a231ed26195R11-R12), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-1e5a8875fbda16cf406f25e6def8d955c7774eeb32da266d24185a231ed26195R53-R55), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-1e5a8875fbda16cf406f25e6def8d955c7774eeb32da266d24185a231ed26195R197-R247), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-1e5a8875fbda16cf406f25e6def8d955c7774eeb32da266d24185a231ed26195R259-R261))
* Add and implement `PatientPrivacyToggle` component for rendering and toggling privacy mode for camera feed (F13, [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-f6f40e1b249471f1821a43cd90bfdce470f563c22b22272a154f9bd7a2de599cR51-R52), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-f6f40e1b249471f1821a43cd90bfdce470f563c22b22272a154f9bd7a2de599cL1128-R1145), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-a0c88ccc2116ceb542bc9be228c54ed98b0bc89e2d9daf66c941f1a0305e626eR331-R340), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-21639e8b3b95469edd34a3d2641a85e6eb02393c442301e11d91665f10f1cf7cR284-R287), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-e2c5c73d556d02503d5adc253828ddf5bad94f6fb9fba25df333d5ca6b9b5c56R130-R134))
* Add and implement `CameraBoundaryConfigure` and `UpdateCameraBoundaryConfigure` components for rendering and performing boundary preset actions for camera feed (F6, [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-1d65a996f952fb6cd0b8721be6e07b3b0bc719c35adb7ddd256b3e550a02a7dbL6-R10), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-1d65a996f952fb6cd0b8721be6e07b3b0bc719c35adb7ddd256b3e550a02a7dbL45-R133), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-1d65a996f952fb6cd0b8721be6e07b3b0bc719c35adb7ddd256b3e550a02a7dbR159-R189), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-1d65a996f952fb6cd0b8721be6e07b3b0bc719c35adb7ddd256b3e550a02a7dbL104-R329), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-1d65a996f952fb6cd0b8721be6e07b3b0bc719c35adb7ddd256b3e550a02a7dbL129-L131), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-1d65a996f952fb6cd0b8721be6e07b3b0bc719c35adb7ddd256b3e550a02a7dbR357-R358), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-1d65a996f952fb6cd0b8721be6e07b3b0bc719c35adb7ddd256b3e550a02a7dbL144-R366), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-1d65a996f952fb6cd0b8721be6e07b3b0bc719c35adb7ddd256b3e550a02a7dbL202-R446))
* Modify `CameraConfigure` component for rendering and performing preset actions for camera feed (F7, [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-1d65a996f952fb6cd0b8721be6e07b3b0bc719c35adb7ddd256b3e550a02a7dbL13))
* Modify `ONVIFCamera` component for rendering and passing props to `LiveFeed` component ([link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-1d65a996f952fb6cd0b8721be6e07b3b0bc719c35adb7ddd256b3e550a02a7dbR21-R23), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-1d65a996f952fb6cd0b8721be6e07b3b0bc719c35adb7ddd256b3e550a02a7dbL202-R446))
* Modify `SwitchTabs` component to accept custom class name prop ([link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-496c84feeecc2c8878e77fdf0fbb641b0e484f9dd292ce0dbe20222f6b5d46b4R1), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-496c84feeecc2c8878e77fdf0fbb641b0e484f9dd292ce0dbe20222f6b5d46b4L7-R15))
* Modify `FileUpload` component to improve voice recorder layout ([link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-6624210f1e9cdff909251ba965fafd78918e3d0a95fb569b8f151e8528eca8b6L1460-R1460))
* Modify `ConsultationDetails` component to wrap camera feed tab content in `Page` component ([link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-f6f40e1b249471f1821a43cd90bfdce470f563c22b22272a154f9bd7a2de599cR51-R52), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-f6f40e1b249471f1821a43cd90bfdce470f563c22b22272a154f9bd7a2de599cL1128-R1145))
* Modify `BedModel` and `CurrentBed` interfaces to add `meta` and `privacy` properties ([link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-e2841f01e53119db63c01994d866100758f06ecc85d11914eecc7cdfe8a2f4dfR201), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-e2841f01e53119db63c01994d866100758f06ecc85d11914eecc7cdfe8a2f4dfR214))
* Add `direction` and `BoundaryRange` types to `src/Common/constants.tsx` ([link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-37705b30476b14631124ce42dd5c3500d43fbede16f0e91f9ae2905e4adc6376R1003-R1010))
* Add end-to-end tests for camera boundary and patient privacy features ([link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-d64d0fb61612e0c542cd4d30013e35a4e1227e00ff0fe8bbb3bcfeec0f35223bR1-R70), [link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-f7a5565ec5d72bfc24a2a4bdccdd257b25c6b364314d9ef17f5c2f527d44b0caR1-R23))
* Add space character to `if` statement in `src/index.tsx` ([link](https://github.com/coronasafe/care_fe/pull/6223/files?diff=unified&w=0#diff-0b5adbfe7b36e4ae2f479291e20152e33e940f7f265162d77f40f6bdb5da7405L12-R12))

### CODE REVIEW

**Comment by JahnabDutta** [comment]
@nihal467 Hey, I have raised this PR for testing and review.

**Comment by nihal467** [comment]
@JahnabDutta the cypress is failed, fix it

**Comment by JahnabDutta** [comment]
> @JahnabDutta the cypress is failed, fix it

Hey, I fixed the cypress test related to patient privacy. But for the camera boundary test, we will require a functional camera for the test to succeed. Can we skip this for the time being? @nihal467

**Comment by JahnabDutta** [comment]
Hey, @nihal467 The patient_pricacy test is running fine in my local. I am using the same data present in the dummy-data backend. I am unsure as to why it is failing here.
image

**Comment by nihal467** [comment]
@JahnabDutta 
image

- the dev camera is not part of dummy data, use any camera named "Dummy camera"

**Comment by nihal467** [comment]
**Initial Review of QA pointers**

**Works Good**

- Camera Feed privacy is working properly
- Multiple user camera access control working good

**issues needed to be addressed**

- Camera Boundary preview is not working, even though in the network tab, the requests are coming in, and the camera is not moving to show the boundary of the camera view.
- Camera boundary preset is hard to set, as it keeps on rotating to a default preset
- When I am moving the camera, to set a boundary, on each click I get a notification, so change it to show notification only when we set a preset of left, right, or any one side and move to the next one.

image

- The UI shows multiple users and the way the subscribe button is kind of not giving a good user experience.
- When the second user is trying to move the camera, even though the camera is not moving as expected but, there is no notification showing the camera is already in use, request access to gain control, or something like that
- The cypress is also failing

Note: Prioritise and work on the issues accordingly, feel free to reach out for any queries

File: src/style/index.css [hunk 1]
```diff
@@ -400,6 +404,47 @@ button:disabled,
...
+}
+
+.bottom-border-flash {
+    animation: borderPulse 1s infinite ease-out;
+    border-bottom: 10px solid #ffffff;
+    border-radius: 20px;
+}
+
+
+@keyframes borderPulse {
```
Comment (khavinshankar) [inline src/style/index.css:443]: just a minor suggestion, instead of camel case, use snake case with hyphen

```suggestion
@keyframes border-pulse {
```

also change it in the places this is used

File: src/Common/hooks/useNotificationSubscribe.ts [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+import { useState } from "react";
+import * as Sentry from "@sentry/browser";
+import { useTranslation } from "react-i18next";
+import { Error } from "../../Utils/Notifications.js";
+import useAuthUser from "../../Common/hooks/useAuthUser";
+
+export default function useNotificationSubscribe() {
+  const { username } = useAuthUser();
+  const dispatch: any = useDispatch();
+  const [isSubscribed, setIsSubscribed] = useState("");
```
Comment (khavinshankar) [inline src/Common/hooks/useNotificationSubscribe.ts]: The naming seems to be bit off, why not change it to some thing like `subscriptionStatus`. 

Also add `type subscriptionStatus = "NotSubscribed" | "SubscribedOnAnotherDevice" | ...` instead of just having generic string type for state

File: src/Common/hooks/useNotificationSubscribe.ts [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        setIsSubscribed("NotSubscribed");
+      } else if (subscription?.endpoint === res?.data?.pf_endpoint) {
+        setIsSubscribed("SubscribedOnThisDevice");
+      } else {
+        setIsSubscribed("SubscribedOnAnotherDevice");
+      }
+    } catch (error) {
+      Sentry.captureException(error);
+    }
+  };
```
Comment (khavinshankar) [inline src/Common/hooks/useNotificationSubscribe.ts]: I assume we are exporting this function and using it inside use effect in the importing file, Why not directly do it here?

File: src/Common/hooks/useNotificationSubscribe.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+      setIsSubscribed("SubscribedOnThisDevice");
+    }
+    setIsSubscribing(false);
+  }
+
+  return {
+    isSubscribed,
+    isSubscribing,
+    setIsSubscribed,
+    setIsSubscribing,
```
Comment (khavinshankar) [inline src/Common/hooks/useNotificationSubscribe.ts]: do we need to export these, Can we try to handle all the state changes through helper functions like subscribe and unsubscribe and so on?

File: src/Components/Assets/AssetType/ONVIFCamera.tsx [hunk 5]
```diff
@@ -42,8 +47,90 @@ const ONVIFCamera = (props: ONVIFCameraProps) => {
   const [refreshPresetsHash, setRefreshPresetsHash] = useState(
     Number(new Date())
   );
+  const [boundaryPreset, setBoundaryPreset] = useState<any>(null);
+  const [toUpdateBoundary, setToUpdateBoundary] = useState<boolean>(false);
+  const [loadingAddBoundaryPreset, setLoadingAddBoundaryPreset] =
+    useState<boolean>(false);
+  const [presets, setPresets] = useState<any[]>([]);
```
Comment (khavinshankar) [inline src/Components/Assets/AssetType/ONVIFCamera.tsx:54]: instead of any, do add types, same for boundary preset

File: src/Components/Common/components/SwitchTabs.tsx [hunk 6]
```diff
@@ -1,12 +1,18 @@
...
   Tab2: string;
+  className?: string;
 }) {
+  const className = classNames(
+    props.className,
+    "relative grid w-full grid-cols-2 items-center gap-4 rounded-md bg-primary-500/10 px-4 py-3 lg:w-52"
+  );
   return (
-    <div className="relative grid w-full grid-cols-2 items-center gap-4 rounded-md bg-primary-500/10 px-4 py-3 lg:w-52">
+    <div className={className}>
```
Comment (khavinshankar) [inline src/Components/Common/components/SwitchTabs.tsx]: why separate it, as it is not reused why not do,

```js
<div className={classNames(
    props.className,
    "relative grid w-full grid-cols-2 items-center gap-4 rounded-md bg-primary-500/10 px-4 py-3 lg:w-52"
  )}>
```

File: src/Components/Patient/PatientPrivacyToggle.tsx [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+      return true;
+
+    return false;
+  };
+  //hook to fetch the privacy info of the patient
+  useEffect(() => {
+    const getPrivacyInfo = async () => {
+      if (
+        consultation?.current_bed?.privacy == true ||
+        consultation?.current_bed?.privacy == false
```
Comment (khavinshankar) [inline src/Components/Patient/PatientPrivacyToggle.tsx:41]: what are we checking here, why are you doing both true and false?

File: src/Components/Patient/PatientPrivacyToggle.tsx [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+            consultationId,
+            bedId as string,
+            consultationBedID as string
+          )
+        );
+        if (
+          res &&
+          res.status === 200 &&
+          res?.data &&
+          (res.data?.privacy == true || res.data?.privacy == false)
```
Comment (khavinshankar) [inline src/Components/Patient/PatientPrivacyToggle.tsx]: same here, if you are checking weather the ...privacy is boolean or not use typeof ...privacy === "boolean"

Also you are using == which means loose comparison, so the above check is always true

File: src/Components/Patient/PatientPrivacyToggle.tsx [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+            <CareIcon className="care-l-unlock text-3xl" />
+            <span className="tooltip-text tooltip-bottom -translate-x-1/2 text-sm">
+              Unlock Privacy
+            </span>
+          </button>
+        )}
+      </div>
+    );
+
+  return <></>;
```
Comment (khavinshankar) [inline src/Components/Patient/PatientPrivacyToggle.tsx]: negate the condition and revert the returns, will be easy to read

**Review by khavinshankar — COMMENTED** [review]
LGTM, some minor changes required

Will re-review after QA is done.

cc: @nihal467

**Comment by khavinshankar** [comment]
ETD: 15th November 

cc: @nihal467

**Comment by khavinshankar** [comment]
@nihal467 Can you test this pr and list out all the issues (separate them wrt priority)

**Comment by nihal467** [comment]
@khavinshankar can you clear merge conflict

**Comment by nihal467** [comment]
https://github.com/coronasafe/care/pull/1518
https://github.com/coronasafe/care/pull/1467
https://github.com/coronasafe/care/pull/1453

@khavinshankar can you fix this backend and create a single PR for testing

File: src/Common/hooks/useNotificationSubscribe.ts (hunk 2 above)
Comment (Pranshu1902) [reply src/Common/hooks/useNotificationSubscribe.ts]: done

File: src/Components/Patient/PatientPrivacyToggle.tsx (hunk 7 above)
Comment (Pranshu1902) [reply src/Components/Patient/PatientPrivacyToggle.tsx:41]: I guess he is trying to check if the value exists for that variable

File: src/Components/Patient/PatientPrivacyToggle.tsx (hunk 9 above)
Comment (Pranshu1902) [reply src/Components/Patient/PatientPrivacyToggle.tsx]: done

**Comment by khavinshankar** [comment]
@Pranshu1902 can you resolve the merge conflicts and wire up the respective backend for this, so that @nihal467 can test it

**Comment by Pranshu1902** [comment]
> [coronasafe/care#1518](https://github.com/coronasafe/care/pull/1518) [coronasafe/care#1467](https://github.com/coronasafe/care/pull/1467) [coronasafe/care#1453](https://github.com/coronasafe/care/pull/1453)
> 
> @khavinshankar can you fix this backend and create a single PR for testing

Combined Single PR for backend: [#1757](https://github.com/coronasafe/care/pull/1757)

**Comment by Pranshu1902** [comment]
> @Pranshu1902 can you resolve the merge conflicts and wire up the respective backend for this, so that @nihal467 can test it

Done
@nihal467 the backend PR is merged into a single PR. You can test this PR

File: src/Redux/actions.tsx [hunk 10]
```diff
@@ -301,6 +301,16 @@ export const updateConsultationBed = (
...
 
+export const togglePatientPrivacy = (external_id: string) =>
+  fireRequest(
+    "togglePatientPrivacy",
+    [],
+    {},
+    {
+      external_id: external_id,
+    }
+  );
```
Comment (rithviknishad) [inline src/Redux/actions.tsx]: It'd be nice to have this converted to use `request` instead of a `fireRequest` action.

(This could be done in a later PR too)

Comment (Pranshu1902) [reply src/Redux/actions.tsx]: let's do this in a later PR, there are some other minor changes as well

**Comment by nihal467** [comment]
@Pranshu1902 
image

- [ ] The above is the SS of the comment I made in the initial testing of this PR, Are all the issues addressed and fixed in the PR?
- [ ] Attached a demo video of the PR,  so to make sure all the functionality are working well, before the QA process starts

**Comment by Pranshu1902** [comment]
@nihal467 Video recording of setting the boundary and previewing it.

https://github.com/coronasafe/care_fe/assets/70687348/d74d7f09-9daf-41bc-8aea-561e3a65c90d

The major reason it failed is because we have a slow network and the camera's view is behind the current time. To check ensure that the time displayed on the top of the camera feed is the same as the current time.

- The notification issue was already resolved by Janab, I verified it
- After setting each boundary the camera is not moving to the default left/right/up/down value and not the default center or something
- Notification when user doesn't have camera move access is also added, user will get a notification saying that the camera is occupied by this user.

**Comment by Pranshu1902** [comment]
Fixed camera boundary cypress tests and updated it to use dummy data for testing
image

**Comment by Pranshu1902** [comment]
Fixed patient privacy cypress test for dummy data
Screenshot from 2023-12-29 15-07-28

@nihal467 both cypress tests will pass once the backend [PR](https://github.com/coronasafe/care/pull/1757) is merged

**Comment by nihal467** [comment]
@Pranshu1902 add a functionality video of the feature covering all the things included in this PR

- [ ] how the camera boundary is configured on asset page
- [ ] how the camera boundary functions on the patient consultation page, and when the user tries to move the camera outside the boundary, how the error is handled
- [ ] how the multiple user data are reflected in the patient consultation page and the request toggle access functionality works, when a second user requests the access
- [ ] how the privacy mode works and verify the functionality of users trying to access the camera when it is on and off

if I miss something, add that too into the QA video and submit it for review.

**Comment by nihal467** [comment]
@Pranshu1902 what is the status on this PR, we need to close it ASAP

**Comment by khavinshankar** [comment]
@Pranshu1902 please fix the cypress tests, it seems only boundary tests are failing

cc: @nihal467

File: src/Common/hooks/useNotificationSubscribe.ts [hunk 11]
```diff
@@ -0,0 +1,138 @@
+import { useDispatch } from "react-redux";
+import {
+  getUserPnconfig,
+  updateUserPnconfig,
+  getPublicKey,
+} from "../../Redux/actions";
```
Comment (rithviknishad) [inline src/Common/hooks/useNotificationSubscribe.ts]: replace to use useQuery/request. Newer codebase should not contain calls with redux.

File: src/Components/Assets/AssetType/ONVIFCamera.tsx [hunk 12]
```diff
@@ -1,15 +1,23 @@
 import { useEffect, useState } from "react";
 import { AssetData, ResolvedMiddleware } from "../AssetTypes";
+import { useDispatch } from "react-redux";
+import {
+  partialUpdateAssetBed,
+  listAssetBeds,
+  deleteAssetBed,
+} from "../../../Redux/actions";
```
Comment (rithviknishad) [inline src/Components/Assets/AssetType/ONVIFCamera.tsx]: replace to useQuery/request

File: src/Components/Facility/Consultations/Feed.tsx [hunk 13]
```diff
@@ -13,6 +13,9 @@ import {
 } from "../../../Common/hooks/useMSEplayer";
 import { PTZState, useFeedPTZ } from "../../../Common/hooks/useFeedPTZ";
 import { useEffect, useRef, useState } from "react";
+import { listAssetBeds } from "../../../Redux/actions";
```
Comment (rithviknishad) [inline src/Components/Facility/Consultations/Feed.tsx]: replace to useQuery/request

File: src/Components/Patient/PatientPrivacyToggle.tsx [hunk 14]
```diff
@@ -0,0 +1,183 @@
+import * as Notification from "../../Utils/Notifications.js";
+import CareIcon from "../../CAREUI/icons/CareIcon.js";
+import { ConsultationModel } from "../Facility/models.js";
+import { useDispatch } from "react-redux";
```
Comment (rithviknishad) [inline src/Components/Patient/PatientPrivacyToggle.tsx]: replace to useQuery/request

File: src/Redux/actions.tsx [hunk 15]
```diff
@@ -44,6 +44,86 @@ export const deleteAssetBed = (asset_id: string) =>
...
+
```
Comment (rithviknishad) [inline src/Redux/actions.tsx]:
```suggestion

```

some of the actions were already migrated. why was it added back?

replace to useQuery/request

File: src/Redux/actions.tsx [hunk 16]
```diff
@@ -258,6 +338,18 @@ export const operateAsset = (id: string, params: object) =>
...
+};
+
+export const getUserPnconfig = (pathParams: object) => {
+  return fireRequest("getUserPnconfig", [], {}, pathParams);
+};
+
+export const updateUserPnconfig = (params: object, pathParams: object) => {
+  return fireRequest("updateUserPnconfig", [], params, pathParams);
+};
+
```
Comment (rithviknishad) [inline src/Redux/actions.tsx]: not required in useQuery/ request


```suggestion
```

File: src/Redux/fireRequest.tsx [hunk 17]
```diff
@@ -131,6 +131,11 @@ export const fireRequest = (
             return;
           }
 
+          // 409 is for conflict on access to an asset
+          if (error.response.status === 409 && key === "operateAsset") {
+            return error.response;
+          }
+
```
Comment (rithviknishad) [inline src/Redux/fireRequest.tsx]: add this to onResponse of the request options after migrating to useQuery/request and remove it from firerequest

**Review by rithviknishad — CHANGES_REQUESTED** [review]

**Comment by nihal467** [comment]
@Pranshu1902 

- [ ] Attach a full QA video of the PR
- [ ] attach the screenshot of passing the cypress test
- [x] fix the backend test

**Comment by gigincg** [comment]
@khavinshankar Can you go through the PR, find out if it's usable or not? And update here?

**Comment by bodhish** [comment]
Closing this PR due to lack of recent progress.

@gigincg @nihal467

**Comment by khavinshankar** [comment]
@bodhish I'm currently working on this PR, I will create new PRs in place of this, lets keep this open till that.

**Comment by khavinshankar** [comment]
Closing this PR, new PR will be raised in its place
